segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -121,9 +119,13 @@
      * Buffer for reading and writing transaction data entries to underlying KV store.
      * This allows lazy storing and avoiding unnecessary load for recently/frequently updated key value pairs.
      */
-    @GuardedBy("lock")
     private final ConcurrentHashMap<String, TransactionData> bufferedTxnData;
 
+    /**
+     * {@link MultiKeyLockManager} instance that manages locks.
+     */
+    private final MultiKeyLockManager lockManager = new MultiKeyLockManager();
```

Comment:
I am not fully convinced that we need both ReadWriteLock and ReentrantReadWriteLock. Why can't you just create a dummy `Object`, map it to a `String` and use `synchronized` on it?

Reply:
1. The bigger issue is that reads block each other (if reading metadata about segment ends up in page read it is a different request on a possibly different thread. This will result in a deadlock. )  If there are multiple parallel read requests on same segment (especially possible with metadata segments) then we don't want the read to block each other.

2.  We'll have to use array of objects and the structured as
```
 synchronized (ob1) {
     synchronized (ob2) {
           synchronized (ob3) {
                      ....  {
                             ....  {
                        // Do work
                           }
                      }
           }
      }
}
```
We can do some lambda magic but really synchronized does not solve the problem for multiple concurrent readers